File: packages/js-dpp/lib/errors/DuplicateIndexError.js
const AbstractIndexError = require('./AbstractIndexError');

class DuplicateIndexError extends AbstractIndexError {
  /**
   * @param {rawContract} rawContract
   * @param {string} documentType
   * @param {Object} indexDefinition
   */
  constructor(rawContract, documentType, indexDefinition) {
    const message = `Duplicate index definition for "${documentType}" document`;

    super(
      message,
      rawContract,
      documentType,
      indexDefinition,
    );
  }
}

module.exports = DuplicateIndexError;

File: packages/rs-dpp/src/errors/consensus/basic/data_contract/duplicate_index_error.rs
use crate::consensus::basic::BasicError;
use crate::consensus::ConsensusError;
use crate::errors::ProtocolError;
use bincode::{Decode, Encode};
use platform_serialization_derive::{PlatformDeserialize, PlatformSerialize};
use thiserror::Error;

#[derive(
    Error, Debug, Clone, PartialEq, Eq, Encode, Decode, PlatformSerialize, PlatformDeserialize,
)]
#[error("Duplicate '{index_name}' index definition for '{document_type}' document")]
#[platform_serialize(unversioned)]
pub struct DuplicateIndexError {
    /*

    DO NOT CHANGE ORDER OF FIELDS WITHOUT INTRODUCING OF NEW VERSION

    */
    document_type: String,
    index_name: String,
}

impl DuplicateIndexError {
    pub fn new(document_type: String, index_name: String) -> Self {
        Self {
            document_type,
            index_name,
        }
    }

    pub fn document_type(&self) -> &str {
        &self.document_type
    }

    pub fn index_name(&self) -> &str {
        &self.index_name
    }
}

impl From<DuplicateIndexError> for ConsensusError {
    fn from(err: DuplicateIndexError) -> Self {
        Self::BasicError(BasicError::DuplicateIndexError(err))
    }
}
